qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -377,6 +408,82 @@ def supported_bases(self):
     def supports_target(self):
         return True
 
+    def __init__(self):
+        super().__init__()
+        self._decomposer_cache = {}
+
```

Comment:
For another PR: I guess we should document what the lifetime of an instance of a plugin class is for the interface - the implementation is that it lives for the lifetime of a pass of the outer `UnitarySynthesis`, which is one run over the DAG, but we should be clear about it so others can cache like this if they want to.  I also wonder if we should consider re-organising the interface such that the free-form configuration dictionary gets passed to the `__init__`, not the `run` (or both).

Reply:
Well it could be more than one run over a dag, it depends specifically how the pass manager is constructed and used. For example, with `transpile()` it could be used multiple times on each iteration of the optimization loop (either in opt level 3 or if the translation method is set to synthesis). This means the plugin instance for that synthesis pass will be called multiple times depending on how many iterations the optimization loop needs. (which is why I added the cache because it could make a noticeable difference in those cases).

That being said I agree, we might need to revisit how we manage the flow of data between the pass and plugins. When we originally designed it plugins were intended to be more of a black box stateless function, you give it a unitary and it gives you back a dagcircuit. But in practice so far I think all the plugins (now with this pr, previously `default` was stateless) have some level of state and/or caching they do out of band. We should open an issue on it to figure out how to evolve the interface to support this.

Reply:
Opened #7843 to track this.